File: src/main/java/me/jangjunha/ftgo/restaurant_service/service/RestaurantService.java
package me.jangjunha.ftgo.restaurant_service.service;

import jakarta.transaction.Transactional;
import me.jangjunha.ftgo.restaurant_service.domain.CreateRestaurantRequest;
import me.jangjunha.ftgo.restaurant_service.domain.Restaurant;
import me.jangjunha.ftgo.restaurant_service.domain.RestaurantRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.UUID;

@Service
@Transactional
public class RestaurantService {
    private final RestaurantRepository restaurantRepository;

    @Autowired
    public RestaurantService(RestaurantRepository restaurantRepository) {
        this.restaurantRepository = restaurantRepository;
    }

    public Restaurant create(CreateRestaurantRequest request) {
        Restaurant restaurant = new Restaurant(request.name, request.menuItems);
        restaurantRepository.save(restaurant);
        return restaurant;
    }

    public Optional<Restaurant> get(UUID restaurantId) {
        return restaurantRepository.findById(restaurantId);
    }

    public Iterable<Restaurant> getAll() {
        return restaurantRepository.findAll();
    }
}

File: ftgo-api-gateway/src/main/kotlin/me/jangjunha/ftgo/apigateway/proxies/RestaurantService.kt
package me.jangjunha.ftgo.apigateway.proxies

import me.jangjunha.ftgo.apigateway.Destinations
import me.jangjunha.ftgo.restaurant_service.api.Restaurant
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import org.springframework.web.reactive.function.client.awaitExchange
import java.net.URI
import java.util.UUID

@Service
class RestaurantService
@Autowired constructor(
    val destinations: Destinations,
    val client: WebClient,
) {
    suspend fun listRestaurants(): List<Restaurant> = client
        .get()
        .uri(URI(destinations.restaurantServiceUrl).resolve("./restaurants/"))
        .awaitExchange { response ->
            when (response.statusCode()) {
                HttpStatus.OK -> response.awaitBody<List<Restaurant>>()
                else -> throw RuntimeException("Cannot retrieve restaurant list")
            }
        }

    suspend fun findRestaurantById(id: UUID): Restaurant = client
        .get()
        .uri(URI(destinations.restaurantServiceUrl).resolve("./restaurants/$id/"))
        .awaitExchange { response ->
            when (response.statusCode()) {
                HttpStatus.OK -> response.awaitBody(Restaurant::class)
                HttpStatus.NOT_FOUND -> throw RestaurantNotFoundException(id)
                else -> throw RuntimeException("Cannot retrieve restaurant")
            }
        }
}
